src/modules/roc_sndio/target_sox/roc_sndio/sox_sink.cpp

```diff
@@ -28,13 +29,17 @@ SoxSink::SoxSink(core::IAllocator& allocator, const Config& config)
         return;
     }
 
-    if (config.frame_size == 0) {
-        roc_log(LogError, "sox sink: frame size is zero");
+    if (config.latency != 0) {
+        roc_log(LogError, "sox sink: setting io latency not supported by sox backend");
         return;
     }
 
-    if (config.latency != 0) {
-        roc_log(LogError, "sox sink: setting io latency not supported by sox backend");
+    channels_ = config.channels;
+    frame_length_ = config.frame_length;
+
+    if (frame_length_ == 0) {
+        roc_log(LogError, "sox sink: frame length is zero");
+        out_signal_.rate = sample_rate();
```

Comment:
Why do we have this line?

```
out_signal_.rate = sample_rate();
```

I think if it will be executed, it will panic (since `valid_` is false here).

Reply:
I think we should have this but it must be moved after `valid_` is true.

Reply:
About why we need this, on line 47, we are assigning out_signal_.rate to config.sample_rate. So I think we should update it accordingly.

Reply:
Why do you think so?

In ctor we set `out_signal_.rate` to the rate from config:

```
out_signal_.rate = config.sample_rate;
```

Then `open_()` passes `out_signal_` to `sox_open_write()`, which tries to use `out_signal_.rate` if it is 1) non-zero 2) supported by device / file.

Then `sox_open_write()` selects some rate (maybe the requested one, and maybe not) and stores the selected rate into `output_->signal.rate`.

`sample_rate()` checks that the device is already opened and returns `output_->signal.rate`.

So, in other words:

* `out_signal_.rate` is the rate which we're requesting from sox
* `output_->signal.rate` and `sample_rate()` is the rate actually selected by sox

So why we may need this assignment?

The naming is not obvious though. Maybe we want to rename some of these fields.

Reply:
> About why we need this, on line 47, we are assigning out_signal_.rate to config.sample_rate. So I think we should update it accordingly.

Didn't see this message. I see. Well, we don't use `out_signal_` in any way except passing it to `sox_open_write()`, so it seems there is no need to update it. Maybe rename it to `requested_signal_params_` or something, to make it clear that it contains *requested* parameters, not actual ones.